src/client/Microsoft.Identity.Client/Internal/Broker/BrokerSilentRequest.cs

```diff
@@ -58,6 +58,12 @@ private async Task<MsalTokenResponse> SendAndVerifyResponseAsync()
 
         internal void CreateRequestParametersForBroker()
         {
+            if (string.IsNullOrEmpty(_silentParameters.LoginHint) && string.IsNullOrEmpty(_silentParameters.Account.Username))
```

Comment:
How can this happen? `AcquireTokenSilent` either takes an IAccount or a LoginHint. Can Account.Username ever be null?

Reply:
Oh, yes, it can be null... https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/7a7b5a1c6d3769740a10afb83184a2c91befc324/src/client/Microsoft.Identity.Client/IAccount.cs#L20

Reply:
What does Android team do here? This experience of throwing an exception is not great - if MSAL were to manage its own cache (instaed of going to the broker), it would be able to get you a token even if IAccount.Username is null. Tokens are indexed by home_account_id, not by username. Is there a way to pass home_account_id to broker?

Reply:
Android bypasses its own cache for silent requests. Also, the broker requires a username to look up the account in the account manager. I dont think there is a way to use the home account id alone

Reply:
We can follow this release up with 4.9.1 to handle this more gracefully

Reply:
I might need to see the end to end of this to understand it, but shouldn't you throw an MsalUiRequiredException here, with the fancy error message, because what/how will the developer address this error? how will they know the login hint or username if there is no user and they are doing a silent call at app start up (like we recommend)...does that make sense?

Reply:
So, if account.Username is null, you cannot perform `AcquireTokenSilent` via broker ever. It might even be better to try to look in MSAL's cache, it's quite possible that you have a token. 

But anyway, we're not designing the feature, I'm just trying to understand what the Android folks are doing - since they designed the feature. Do you know if they also throw / fail if loginHint is null and Username is null?

Reply:
@bgavrilMS , you are correct, the broker can use the "home_account_id" and the "local_account_id" to look for the account, I need to change the conditional in the android broker to look for "home_account_id" and "local_account_id" instead of just the username.

Reply:
should you just be checking for a null account? if the account is null, you can't have the home account id, right? so just check for a null account. in other words, can an account have an environment value, but no username and no home account id?

---
In reply to: [381623173](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1649#discussion_r381623173) [](ancestors = 381623173)

Reply:
Discussed offline, adding comment to better explain the scenario here

Reply:
tracking issue?

---
In reply to: [381614475](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1649#discussion_r381614475) [](ancestors = 381614475)